<a href="https://colab.research.google.com/github/Kevin-AngelQD/padp/blob/main/Actividad_Calificada_07_Datos_Duplicados_271758.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Kevin Angel Quiñonez Dominguez |
| **Matrícula** |271758 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [2]:
# Escribe tu código aquí
print("Filas duplicadas donde no se cuenta la primera aparicion de la fila: ",hoteles.duplicated(keep='first').sum())
print("Filas duplicadas donde no se cuenta la ultima aparicion de la fila: ",hoteles.duplicated(keep='last').sum())
print("Todas las filas que tienen un duplicado: ",hoteles.duplicated(keep=False).sum())

#Como se menciono en el ejercicio de practica first y last siempre dan el mismo conteo, false da todas las filas con duplicados

Filas duplicadas donde no se cuenta la primera aparicion de la fila:  31994
Filas duplicadas donde no se cuenta la ultima aparicion de la fila:  31994
Todas las filas que tienen un duplicado:  40165


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [3]:
# Escribe tu código aquí
print("Porcentaje de filas sobrantes: ",(hoteles.duplicated(keep=False).sum()/hoteles.shape[0])*100,"%")
print("Numero de grupos repetidos: ",hoteles.duplicated(keep=False).sum()-hoteles.duplicated(keep='first').sum())

#La parte afectada representa mas de la tercera parte del dataset


Porcentaje de filas sobrantes:  33.64184605075802 %
Numero de grupos repetidos:  8171


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [4]:
# Escribe tu código aquí
hoteles_copia=hoteles.copy()
hoteles_copia["es_duplicado"] = hoteles.duplicated(keep="first")

total_reservas = hoteles_copia.groupby("hotel")["es_duplicado"].count()
filas_duplicadas = hoteles_copia.groupby("hotel")["es_duplicado"].sum()

# Combine los resultados en un nuevo dataframe
resumen = pd.DataFrame(
    {
        "Total_Reservas": total_reservas,
        "Filas_Duplicadas": filas_duplicadas,
        "Porcentaje_%": (filas_duplicadas / total_reservas) * 100,
    }
).reset_index()

print(resumen)



          hotel  Total_Reservas  Filas_Duplicadas  Porcentaje_%
0    City Hotel           79330             25902     32.650952
1  Resort Hotel           40060              6092     15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

Es considerablemente grave, mas grave en City Hotel que en Resort Hotel. Pero de igual manera no se puede pasar por alto. Resort Hotel tiene poco menos de la mitad de City Hotel.

City Hotel en realidad tiene 53,428 de 79,330 datos limpios. 67.35% de datos utilizables y
Resort Hotel 33,968 de 40,060. 84.8% de datos utilizables

En terminos de diferencia se pierde una cantidad considerable de datos en City Hotel, casi un tercio. En Resort Hotel seria un sexto aproximadamente, es menor problema en Resort Hotel pero de igual manera no se puede ignorar.

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [5]:
# Escribe tu código aquí
#Considero que el tiempo es la mejor forma ya que sirve de manera indirecta como un ID y tomando en cuenta que multiples columnas se maneja parte del tiempo
#se decidio utilizarlas tomando en cuenta el tiempo de reservacion, llegada, dias entre la reserva y hotel si una es igual, no dara problemas ya que hay
#mas variantes que organicen a pesar de compartir muchos tiempo similares.

#Columnas relacionadas al tiempo
col_ord = [
    'reservation_status_date',
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_week_number',
    'arrival_date_day_of_month',
    'lead_time',
    'hotel'
]

#Columnas de muestra
col_muestra = [
    'hotel',
    'lead_time',
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_day_of_month',
    'adults', 'country', 'market_segment',
    'deposit_type', 'adr', 'reservation_status'
]

#print(hoteles_copia[col_muestra].head(12))
#print(hoteles_copia[hoteles_copia.duplicated(keep=False)].sort_values(by=col_ord).head(12))
print((hoteles_copia[hoteles_copia.duplicated(keep=False)].sort_values(by=col_ord))[col_muestra].head(12))





            hotel  lead_time  arrival_date_year arrival_date_month  \
73815  City Hotel        300               2015             August   
73816  City Hotel        300               2015             August   
73817  City Hotel        300               2015             August   
73818  City Hotel        300               2015             August   
73819  City Hotel        300               2015             August   
73820  City Hotel        300               2015             August   
73821  City Hotel        300               2015             August   
73822  City Hotel        300               2015             August   
73823  City Hotel        300               2015             August   
73824  City Hotel        300               2015             August   
73825  City Hotel        300               2015             August   
73826  City Hotel        300               2015             August   

       arrival_date_day_of_month  adults country market_segment deposit_type  \
73815    

### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [6]:
# Escribe tu código aquí

# dropna=True por defecto: Ignora la fila con el nulo
print(hoteles_copia.value_counts().head(10))


print("Combinacion de valores del grupo mas grande: \n",hoteles_copia.value_counts().head(1)) #Son 6 repetidos

hotel         is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type    adr     required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date  es_duplicado
City Hotel    0            256        2016               October             43                        16                         2                        3                     2       0.0       0       BB    DEU      Online TA       TA/TO                 0                  0                       0                               A                   A                   0                No Deposit    9.0    333.0    0  

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [7]:
# Escribe tu código aquí

print("Distribucion de market_segment: \n",hoteles_copia[hoteles_copia.duplicated(keep=False)].market_segment.value_counts(normalize=True))
print("Distribucion de deposit_type: \n",hoteles_copia[hoteles_copia.duplicated(keep=False)].deposit_type.value_counts(normalize=True))
print("Tasa de cancelacion: \n",hoteles_copia[hoteles_copia.duplicated(keep=False)].is_canceled.value_counts(normalize=True))
#Alta metrica de reservas canceladas (puede deberse a tener una reserva duplicada por accidente y se cancelaron o que se duplico la reserva cancelada en si)

Distribucion de market_segment: 
 market_segment
Groups           0.529154
Offline TA/TO    0.346884
Online TA        0.081231
Corporate        0.030675
Direct           0.011536
Complementary    0.000445
Aviation         0.000074
Name: proportion, dtype: float64
Distribucion de deposit_type: 
 deposit_type
No Deposit    0.499407
Non Refund    0.498850
Refundable    0.001743
Name: proportion, dtype: float64
Tasa de cancelacion: 
 is_canceled
1    0.685423
0    0.314577
Name: proportion, dtype: float64


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

a) Considero que desde mi punto de vista seria muy improbable casi imposible que 6 personas hicieran exactamente las misma reservacion al mismo tiempo, mismo dia de llegada, dos adultos, mismo lead_time, incluso mismo tipo de habitacion reservado y entregado, la probabilidad de esto es infinitesimal. Agregando un poco de mi conocimiento en hoteleria, tengo entendido que cada reserva hace que aumente el precio de habitacion, una cantidad que diferenciaria cada reservacion en el ADR, por lo que infiriendo en la actividad 2.2  esos 6 podrian ser error de captura en el sistema ("double-posting" o reintentos de envío tras un congelamiento del sistema)

b) Tomando en cuenta que .duplicated() trata a los valores NaN como identicos entre si puede que haya una sobreestimacion de duplicados reales. Al haber tantos NaN idénticos en estas dos columnas, .duplicated() tiende a empaquetar como "duplicados" a registros que quizás solo se parecen en las variables generales, arrastrados por la coincidencia forzada de los campos vacíos. Una forma de asegurarse sobre los duplicados seria no tomar en cuenta las columnas agent y company para garantizar que la comparacion de filas sea limpia y transparente.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [8]:
# Escribe tu código aquí
hoteles_limpio = hoteles_copia.drop_duplicates(keep='first').reset_index(drop=True)
print("Filas antes: ",hoteles_copia.shape[0])
print("Filas despues: ",hoteles_limpio.shape[0])
print("Filas eliminadas: ",hoteles_copia.shape[0]-hoteles_limpio.shape[0])

print(hoteles_limpio) #Confirmar la renumeracion del indice

Filas antes:  119390
Filas despues:  95567
Filas eliminadas:  23823
              hotel  is_canceled  lead_time  arrival_date_year  \
0      Resort Hotel            0        342               2015   
1      Resort Hotel            0        737               2015   
2      Resort Hotel            0          7               2015   
3      Resort Hotel            0         13               2015   
4      Resort Hotel            0         14               2015   
...             ...          ...        ...                ...   
95562    City Hotel            0         23               2017   
95563    City Hotel            0        102               2017   
95564    City Hotel            0         34               2017   
95565    City Hotel            0        109               2017   
95566    City Hotel            0        205               2017   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0                   July                        27        

### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [9]:
# Escribe tu código aquí
resumen_original = hoteles.groupby("hotel")[['is_canceled', 'adr']].mean().reset_index()
resumen_limpio = hoteles_limpio.groupby("hotel")[['is_canceled', 'adr']].mean().reset_index()

print("Resumen original: \n",resumen_original)
print("Resumen limpio: \n",resumen_limpio)

Resumen original: 
           hotel  is_canceled         adr
0    City Hotel     0.417270  105.304465
1  Resort Hotel     0.277634   94.952930
Resumen limpio: 
           hotel  is_canceled         adr
0    City Hotel     0.311486  109.718843
1  Resort Hotel     0.242857   97.743482


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [10]:
# Escribe tu código aquí

resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'
comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index=True)
comparacion = comparacion.sort_values(by='hotel')
print(comparacion)



          hotel  is_canceled         adr         version
0    City Hotel     0.417270  105.304465        original
2    City Hotel     0.311486  109.718843  sin duplicados
1  Resort Hotel     0.277634   94.952930        original
3  Resort Hotel     0.242857   97.743482  sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

a) El procentaje de cancelacion bajo considerablemente en el City Hotel ese hotel tenia un tercio de datos totales duplicados (como se menciono en la actividad 2) por lo que el eliminar los duplicados redujo la tasa de cancelacion y aumento el adr.

b) El modelo causaria un sobreajuste, memoriza los duplicados en vez de aprender las reglas generales, daria metricas de evaluacion engañosas (como ya vimos en 3.3 y la diferencia considerable de resultados) tambien tendria un mayor costo computacional ya que entrenar un modelo con miles de filas idénticas innecesarias consume más memoria RAM y aumenta el tiempo de procesamiento de los scripts de Python.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [11]:
# Escribe tu código aquí

print("Duplicados exactos: ",hoteles_limpio.duplicated(keep=False).sum())
print("Indices repetidos: ",hoteles_limpio.index.duplicated().sum())
print("Longitud: ",len(hoteles_limpio) - 1)

Duplicados exactos:  0
Indices repetidos:  0
Longitud:  95566


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [12]:
# Escribe tu código aquí
print("Filas sobrantes: ",hoteles_limpio[hoteles_limpio.duplicated(subset=hoteles_limpio.columns.difference(['reservation_status_date']), keep=False)].shape[0])
print("Filas sobrantes: \n",hoteles_limpio[hoteles_limpio.duplicated(subset=hoteles_limpio.columns.difference(['reservation_status_date']), keep=False)][col_muestra])
print("Distribucion de reservation_status: \n",hoteles_limpio[hoteles_limpio.duplicated(subset=hoteles_limpio.columns.difference(['reservation_status_date']), keep=False)].reservation_status.value_counts())

Filas sobrantes:  613
Filas sobrantes: 
               hotel  lead_time  arrival_date_year arrival_date_month  \
2000   Resort Hotel         81               2015          September   
2002   Resort Hotel         81               2015          September   
2887   Resort Hotel         41               2015           November   
2888   Resort Hotel         41               2015           November   
2951   Resort Hotel         36               2015           November   
...             ...        ...                ...                ...   
61277    City Hotel         24               2015           December   
61533    City Hotel         24               2015           December   
61534    City Hotel         24               2015           December   
62041    City Hotel         21               2016            January   
62420    City Hotel         21               2016            January   

       arrival_date_day_of_month  adults country market_segment deposit_type  \
2000          

**Respuesta 4.2:**

Estas filas no pueden ser un duplicado como tal exactamente tomando en cuenta la fecha puede ser que estas sean ajustes de reservacion, en ciertos sistemas no se actualizaria la reservacion y un pequeño detalle o error de dedo puede requerir que hagan otra reservacion y se queda la reservacion con error de dedo. Otra cosa (que considero mas posible siendo realista) puede ser es que estas reservaciones fueran apartadas por una agencia de viaje y por ende dejan ciertas caracteristicas iguales para vender a turistas/pareja, etc, llego el dia de corte, no se vendieron por lo que la agencia la cancela todas de golpe. Estos registros no los eliminaria ya que representan cancelaciones legítimas de bloques o contratos de agencias, y borrarlos subestimaría drásticamente el impacto real de las cancelaciones en el inventario del hotel.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

- **Criterio** de duplicado utilizado (columnas comparadas):
Se aplicó un criterio de coincidencia absoluta en la totalidad de las columnas del dataset (32 variables originales, incluyendo métricas clave como lead_time, arrival_date, adr e is_canceled). Solo se consideraron duplicados aquellos registros donde cada uno de sus campos era 100% idéntico a otro.

- **Regla de conservación** (`keep`):
Se utilizó el parámetro por defecto keep='first' (conservar la primera ocurrencia). En este caso particular, la elección de keep no es relevante ni afecta los resultados debido a que los registros duplicados son idénticos en absolutamente todas sus dimensiones (incluyendo marcas de tiempo implícitas e índices de carga). Al no existir un identificador único (ID de transacción o timestamp de inserción en segundos) que permita priorizar un registro sobre otro, cualquier regla de conservación (first, last) produce el mismo DataFrame final

- **Resultado**:



*   Filas iniciales (antes): 119,390
*   Filas finales (después): 95,567
*   Filas eliminadas: 23,823 (representando el 19.95% del dataset original)


- **Supuestos y limitaciones**:

**Supuesto principal:** Se asume que las 23,823 filas eliminadas corresponden a errores técnicos de persistencia de datos o fallos de red (double-posting por reintentos de los sistemas integrados de reservas). Esto se respalda bajo la lógica del negocio hotelero, donde la probabilidad de que decenas de miles de clientes individuales coincidan al mismo centavo en tarifas dinámicas (adr) y en más de 30 variables independientes es infinitesimal.

**Limitaciones identificadas:** Al no contar con un identificador único de reserva (booking_id), existe el riesgo residual de haber eliminado registros de transacciones reales que compartían características idénticas.

**Recomendación de dominio:** Se advierte que este criterio de limpieza no debe extenderse a filas que solo se parecen parcialmente (como aquellas que varían únicamente en la fecha de reservación y están canceladas). Confundir patrones legítimos del negocio (como las cancelaciones en bloque de agencias) con errores del sistema constituye una limitación de contexto que sesgaría gravemente el análisis comercial.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1.   Si se nos hiciera saber que las habitaciones duplicadas son parte de contrato especial por agencias de viaje, paquetes de turismo, etc, no se deberian eliminar ya que practicamente cada habitacion son ingresos reales que pueden afectar el resultado si por ejemplo se esta analizando los ingresos del hotel.
2. Cualquier informacion que pueda usarse para diferenciar las filas, por ejemplo, nombre y apellido del huesped que reserva. Un id de booking, numero de habitacion preasignada o un timestamp con segundos exactos, esto ayudaria mucho con la incertidumbre de si eran datos duplicados, cualquiera de estos datos puede apoyar en la toma de decisiones.
3.   Puede seguir existiendo duplicados pero no duplicados identicos por ejemplo filas con errores de dedo como puede ser que en la columna market_segment tenga "Groupss" o "Online RA" en vez de "Groups" Y "Online TA" respectivamente. tambien filas casi identicas con datos incompletos una por ejemplo con datos en agent y otra con NaN.




---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.